commit 04cc1e6a6a85c4d09b487b9e29ebaf57c5ba932f
Author: Ian Campbell <ijc@hellion.org.uk>
Date:   Thu Sep 2 14:36:40 2010 +0100

    xen: netback: save interrupt state in add_to_net_schedule_list_tail
    
    add_to_net_schedule_list_tail is called from both hard interrupt context
    (add_to_net_schedule_list_tail) and soft interrupt/process context
    (netif_schedule_work) so use the interrupt state saving spinlock
    variants.
    
    Fixes:
            ------------[ cut here ]------------
            WARNING: at kernel/lockdep.c:2323 trace_hardirqs_on_caller+0xef/0x1a0()
            Hardware name: PowerEdge 860
            Modules linked in: rtc_cmos rtc_core rtc_lib
            Pid: 16, comm: xenwatch Not tainted 2.6.32.18-x86_32p-xen0-00850-ge6b9b2c #98
            Call Trace:
             [<c103951c>] warn_slowpath_common+0x6c/0xc0
             [<c1039585>] warn_slowpath_null+0x15/0x20
             [<c105f60f>] trace_hardirqs_on_caller+0xef/0x1a0
             [<c105f6cb>] trace_hardirqs_on+0xb/0x10
             [<c136cc72>] _spin_unlock_irq+0x22/0x40
             [<c11ab9ef>] add_to_net_schedule_list_tail+0x5f/0xb0
             [<c11aba6b>] netif_be_int+0x2b/0x120
             [<c106dd8e>] handle_IRQ_event+0x2e/0xe0
             [<c106f98e>] handle_level_irq+0x6e/0xf0
             [<c1197cdf>] __xen_evtchn_do_upcall+0x16f/0x190
             [<c11981b8>] xen_evtchn_do_upcall+0x28/0x40
             [<c100b487>] xen_do_upcall+0x7/0xc
             [<c119bcf9>] xs_talkv+0x59/0x1a0
             [<c119bf6a>] xs_single+0x3a/0x50
             [<c119c6f9>] xenbus_read+0x39/0x60
             [<c11adf77>] frontend_changed+0x3e7/0x6a0
             [<c119d35a>] xenbus_otherend_changed+0x8a/0xa0
             [<c119d572>] frontend_changed+0x12/0x20
             [<c119b9dc>] xenwatch_thread+0x7c/0x140
             [<c104ea74>] kthread+0x74/0x80
             [<c100b433>] kernel_thread_helper+0x7/0x10
            ---[ end trace 48d73949a8e0909a ]---
    
    Signed-off-by: Ian Campbell <ian.campbell@citrix.com>
    Signed-off-by: Jeremy Fitzhardinge <jeremy.fitzhardinge@citrix.com>

---
 drivers/xen/netback/netback.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

Index: linux-2.6.32.hg/drivers/xen/netback/netback.c
===================================================================
--- linux-2.6.32.hg.orig/drivers/xen/netback/netback.c	2010-10-07 18:30:03.462905888 +0100
+++ linux-2.6.32.hg/drivers/xen/netback/netback.c	2010-10-07 18:30:08.590902801 +0100
@@ -848,17 +848,19 @@ static void remove_from_net_schedule_lis
 
 static void add_to_net_schedule_list_tail(struct xen_netif *netif)
 {
+	unsigned long flags;
+
 	struct xen_netbk *netbk = &xen_netbk[netif->group];
 	if (__on_net_schedule_list(netif))
 		return;
 
-	spin_lock_irq(&netbk->net_schedule_list_lock);
+	spin_lock_irqsave(&netbk->net_schedule_list_lock, flags);
 	if (!__on_net_schedule_list(netif) &&
 	    likely(netif_schedulable(netif))) {
 		list_add_tail(&netif->list, &netbk->net_schedule_list);
 		netif_get(netif);
 	}
-	spin_unlock_irq(&netbk->net_schedule_list_lock);
+	spin_unlock_irqrestore(&netbk->net_schedule_list_lock, flags);
 }
 
 void netif_schedule_work(struct xen_netif *netif)
